src/op.rs

```diff
@@ -0,0 +1,2691 @@
+    Bytes {
+        /// The value.
+        value: &'input [u8],
+    },
+}
+
+/// The description of a single piece of the result of a DWARF
+/// expression.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Piece<'input> {
+    /// If given, the size of the piece in bits.  If `None`, then the
+    /// piece takes its size from the enclosed location.
+    pub size_in_bits: Option<u64>,
+    /// If given, the bit offset of the piece.  If `None`, then the
+    /// piece starts at the next byte boundary.
+    pub bit_offset: Option<u64>,
+    /// Where this piece is to be found.
+    pub location: Location<'input>,
+}
+
+// A helper function to handle branch offsets.
+fn compute_pc<'input, Endian>(pc: EndianBuf<'input, Endian>,
+                              bytecode: &'input [u8],
+                              offset: i16)
+                              -> ParseResult<EndianBuf<'input, Endian>>
+    where Endian: Endianity
+{
+    let pcbytes: &[u8] = pc.into();
+    let this_len = pcbytes.len();
+    let full_len = bytecode.len();
+    let new_pc = (full_len - this_len).wrapping_add(offset as usize);
+    if new_pc > full_len {
+        Err(Error::BadBranchTarget(new_pc))
+    } else {
+        Ok(EndianBuf::new(&bytecode[new_pc..]))
+    }
+}
+
+#[test]
+fn test_compute_pc() {
+    // Contents don't matter for this test, just length.
+    let bytes = [0, 1, 2, 3, 4];
+    let bytecode = &bytes[..];
+    let ebuf = EndianBuf::<LittleEndian>::new(bytecode);
+
+    match compute_pc(ebuf, bytecode, 0) {
+        Ok(val) => {
+            let valbytes: &[u8] = val.into();
+            assert_eq!(valbytes.len(), bytecode.len());
+        }
+        otherwise => panic!("Unexpected result: {:?}", otherwise),
+    }
+
+    match compute_pc(ebuf, bytecode, -1) {
+        Err(Error::BadBranchTarget(_)) => {}
+        otherwise => panic!("Unexpected result: {:?}", otherwise),
+    }
+
+    match compute_pc(ebuf, bytecode, 5) {
+        Ok(val) => {
+            let valbytes: &[u8] = val.into();
+            assert_eq!(valbytes.len(), 0);
+        }
+        otherwise => panic!("Unexpected result: {:?}", otherwise),
+    }
+
+    match compute_pc(EndianBuf::<LittleEndian>::new(&bytes[3..]), bytecode, -2) {
+        Ok(val) => {
+            let valbytes: &[u8] = val.into();
+            assert_eq!(valbytes.len(), bytecode.len() - 1);
+        }
+        otherwise => panic!("Unexpected result: {:?}", otherwise),
+    }
+
+    match compute_pc(EndianBuf::<LittleEndian>::new(&bytes[2..]), bytecode, 2) {
+        Ok(val) => {
+            let valbytes: &[u8] = val.into();
+            assert_eq!(valbytes.len(), bytecode.len() - 4);
+        }
+        otherwise => panic!("Unexpected result: {:?}", otherwise),
+    }
+}
+
+// Like parse_u8 but takes and returns an EndianBuf for convenience.
+fn parse_u8e<'input, Endian>(bytes: EndianBuf<'input, Endian>)
+                             -> ParseResult<(EndianBuf<'input, Endian>, u8)>
```

Comment:
These parse functions should probably be with the similar functions in parser.rs.  We still need to split up parser.rs a bit though, I'm fine with this for now.